### PR TITLE
[charts] Add context to axis value formatter

### PR DESCRIPTION
**Pull request by alexfauquette**
Part of #10928

### CODE REVIEW

File: packages/x-charts/src/models/axis.ts [hunk 1]
```diff
@@ -209,7 +209,26 @@ export type AxisConfig<S extends ScaleName = ScaleName, V = any> = {
...
+   * @returns {string} The string to display.
+   */
   valueFormatter?: (value: V) => string;
+  /**
+   * Formats the axis value to be displayed in axis ticks.
+   * If defined, this formatter overrides the `valueFormatter` for ticks.
+   * @param {V} value The value to format.
+   * @returns {string} The string to display.
+   */
+  tickValueFormatter?: (value: V) => string;
```
Comment (alexfauquette) [inline packages/x-charts/src/models/axis.ts:224]: I'm hesitating with 
- `tickFormatter` which is shorter and should not interfere with others
- `valueTickFormatter` which has the advantage to group all the formatters at the same place in the alphabetical order

Comment (LukasTy) [reply packages/x-charts/src/models/axis.ts:224]: This is an interesting situation. 🤔 
`tickValueFormatter` seems like it makes the most sense, but have you considered going the other way - adding a context to the existing `valueFormatter`? 🤔 

```ts
interface ValueFomatterContext {
  location: 'value' | 'tick' | 'tooltip';
  // or
  location?: 'tick' | 'tooltip';
}
///
valueFormatter?: (value: V, context: ValueFomatterContext) => string;
```

Comment (alexfauquette) [reply packages/x-charts/src/models/axis.ts:224]: I did not had preferences between the two, but effectively it fixes the naming issue :+1:

File: docs/data/charts/axis/axis.md [hunk 2]
```diff
@@ -59,6 +59,18 @@ Some series types also require specific axis attributes:
 - line plots require an `xAxis` to have `data` provided
 - bar plots require an `xAxis` with `scaleType='band'` and some `data` provided.
 
+### Axis formatter
+
+Axis data can be displayed in the axes ticks, and the in the tooltip.
```
Comment (LukasTy) [inline docs/data/charts/axis/axis.md:64]:
```suggestion
Axis data can be displayed in the axes ticks and the in the tooltip.
```

File: docs/data/charts/axis/FormatterDemoNoSnap.tsx [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+  return (
+    <BarChart
+      dataset={dataset}
+      xAxis={[
+        {
+          scaleType: 'band',
+          dataKey: 'month',
+          valueFormatter: (month, context) =>
+            context.location === 'tick'
+              ? `${month.slice(0, 3)}. \n2023`
```
Comment (LukasTy) [inline docs/data/charts/axis/FormatterDemoNoSnap.tsx:115]: Nitpick: I'd consider skipping the `.` to provide a bit cleaner vertical alignment 🙈 
```suggestion
              ? `${month.slice(0, 3)} \n2023`
```

**Review by LukasTy — APPROVED** [review]
Great work! 👍 💯 
Given the pivot in the solution, I took the liberty to update the PR title.
Feel free to update it if you can think of a more appropriate wording. 👌

File: docs/data/charts/axis/axis.md (hunk 2 above)
Comment (alexfauquette) [reply docs/data/charts/axis/axis.md:64]:
```suggestion
Axis data can be displayed in the axes ticks and the tooltip.
```

File: docs/data/charts/axis/FormatterDemoNoSnap.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+  return (
+    <BarChart
+      dataset={dataset}
+      xAxis={[
+        {
+          scaleType: 'band',
+          dataKey: 'month',
+          valueFormatter: (month, context) =>
+            context.location === 'tick'
+              ? `${month.slice(0, 3)}. \n2023`
```
Comment (alexfauquette) [inline docs/data/charts/axis/FormatterDemoNoSnap.js:115]:
```suggestion
              ? `${month.slice(0, 3)} \n2023`
```

File: docs/data/charts/axis/FormatterDemoNoSnap.tsx.preview [hunk 5]
```diff
@@ -0,0 +1,15 @@
+<BarChart
+  dataset={dataset}
+  xAxis={[
+    {
+      scaleType: 'band',
+      dataKey: 'month',
+      valueFormatter: (month, context) =>
+        context.location === 'tick'
+          ? `${month.slice(0, 3)}. \n2023`
```
Comment (alexfauquette) [inline docs/data/charts/axis/FormatterDemoNoSnap.tsx.preview:9]:
```suggestion
          ? `${month.slice(0, 3)} \n2023`
```

**Pull request merged**